macros/src/lib.rs

```diff
@@ -0,0 +1,98 @@
+// This file is part of radicle-link
+// <https://github.com/radicle-dev/radicle-link>
+//
+// Copyright (C) 2019-2020 The Radicle Team <dev@radicle.xyz>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 or
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+#![feature(proc_macro_diagnostic)]
+
+use std::convert::TryFrom;
+
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, LitStr};
+
+use radicle_git_ext::reference::name::{RefLike, RefspecPattern};
+
+/// Create `RefLike` from a string literal.
+///
+/// The string is validated at compile time, and an unsafe conversion is
+/// emitted.
+///
+/// ```rust
+/// use std::convert::TryFrom;
+/// use radicle_macros::reflike;
+///
+/// assert_eq!("lolek/bolek", reflike!("lolek/bolek").as_str())
+/// ```
+#[proc_macro]
+pub fn reflike(input: TokenStream) -> TokenStream {
+    let lit = parse_macro_input!(input as LitStr);
+
+    match RefLike::try_from(lit.value()) {
+        Ok(safe) => {
+            let safe_str = safe.as_str();
+            let expand = quote! {
+                ::radicle_git_ext::reference::name::RefLike::try_from(#safe_str).unwrap()
+            };
+
+            TokenStream::from(expand)
+        },
+
+        Err(e) => {
+            lit.span()
+                .unwrap()
+                .error(format!("invalid RefLike literal: {}", e))
+                .emit();
+
+            TokenStream::from(quote! { unimplemented!() })
```

Comment:
This is a bit funky -- `rustc` won't stop reporting errors after we emitted ours. If we just emit an empty `TokenStream`, there will be the unhelpful message "expected expression, found end of macro arguments". This way, a series of unused import warnings, and perhaps a type inference error will follow.

Reply:
Could we use `std::compile_error` instead possibly?

Reply:
That's worse -- the span (the wobbly line indicating where the malformed input was) we just emitted will be lost, and instead we'll get a wobbly line below `compile_error`, which takes only a `&'static str`.

Reply:
Fair enough, it was a shot in the dark :sweat_smile: I'll take a closer look now :monocle_face: